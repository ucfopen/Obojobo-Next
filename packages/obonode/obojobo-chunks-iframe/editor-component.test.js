import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'

import IFrame from './editor-component'

import ModalUtil from 'obojobo-document-engine/src/scripts/common/util/modal-util'
jest.mock('obojobo-document-engine/src/scripts/common/util/modal-util')
jest.mock('obojobo-document-engine/src/scripts/oboeditor/util/freeze-unfreeze-editor')
import { Transforms } from 'slate'
jest.mock('slate')
jest.mock('slate-react')
jest.mock(
	'obojobo-document-engine/src/scripts/oboeditor/components/node/with-slate-wrapper',
	() => item => item
)
jest.mock(
	'obojobo-document-engine/src/scripts/oboeditor/components/node/editor-component',
	() => props => <div>{props.children}</div>
)

describe('IFrame Editor Node', () => {
	beforeEach(() => {
		jest.restoreAllMocks()
		jest.resetAllMocks()
	})

	test('IFrame component', () => {
		const component = renderer.create(
			<IFrame
				element={{
					content: {
						width: 200,
						height: 200,
						controls: '',
						border: false,
						initialZoom: 1,
						src: 'mockSrc',
						title: 'mockTitle'
					}
				}}
				parent={{
					getPath: () => ({
						get: () => 0
					}),
					nodes: {
						size: 2
					}
				}}
			/>
		)
		const tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})

	test('IFrame renders with no size correctly', () => {
		const component = renderer.create(
			<IFrame
				element={{
					content: {
						controls: '',
						border: false,
						initialZoom: 1,
						src: 'mockSrc'
					}
				}}
				selected
				parent={{
					getPath: () => ({
						get: () => 0
					}),
					nodes: {
						size: 2
					}
				}}
			/>
		)
		expect(component.toJSON()).toMatchSnapshot()
	})

	test('IFrame renders with long src and no title correctly', () => {
		const component = renderer.create(
			<IFrame
				element={{
					content: {
						width: 200,
						height: 200,
						controls: '',
						border: false,
						initialZoom: 1,
						src: 'mockVeryLongSrcThatHasManyCharsAndExceedsTheMaxDisplaySize'
					}
				}}
				selected
				parent={{
					getPath: () => ({
						get: () => 0
					}),
					nodes: {
						size: 2
					}
				}}
			/>
		)

		const title = component.root.findByType('span')
		expect(title.children[0]).toBe('mockVeryLongSrcThatHasManyCharsAndExceedsTheMaxDis...')
	})

	test('IFrame component deletes self', () => {
		const component = mount(
			<IFrame
				element={{
					content: {
						controls: '',
						border: false,
						initialZoom: 1,
						src: ''
					}
				}}
				parent={{
					getPath: () => ({
						get: () => 0
					}),
					nodes: {
						size: 2
					}
				}}
				editor={{}}
			/>
		)

		component
			.find('button')
			.at(0)
			.simulate('click')

		expect(Transforms.removeNodes).toHaveBeenCalled()
	})

	test('IFrame component opens modals accordingly', () => {
		const component = mount(
			<IFrame
				element={{
					content: {
						controls: '',
						border: false,
						initialZoom: 1,
						src: ''
					}
				}}
				parent={{
					getPath: () => ({
						get: () => 0
					}),
					nodes: {
						size: 2
					}
				}}
			/>
		)

		// Testing opening NewIFrameModal
		component
			.find('button')
			.at(1)
			.simulate('click')
		expect(ModalUtil.show).toHaveBeenCalled()

		// Testing opening EditIFrameModal
		component.instance().openEditIframeModal({
			src: 'mock-src',
			srcFormatted: 'mock-formatted-src'
		})
		expect(ModalUtil.show).toHaveBeenCalled()
	})

	test('IFrame component handles tab', () => {
		const component = mount(
			<IFrame
				element={{
					content: {
						controls: '',
						border: false,
						initialZoom: 1,
						src: ''
					}
				}}
			/>
		)

		component
			.find('button')
			.at(0)
			.simulate('keyDown', { key: 'k' })
		component
			.find('button')
			.at(0)
			.simulate('keyDown', { key: 'Tab', shiftKey: 'true' })

		component
			.find('button')
			.at(1)
			.simulate('keyDown', { key: 'k' })
		component
			.find('button')
			.at(1)
			.simulate('keyDown', { key: 'Tab' })

		const tree = component.html()
		expect(tree).toMatchSnapshot()
	})

	test('changeProperties sets the nodes content', () => {
		const component = mount(
			<IFrame
				element={{
					content: {
						controls: '',
						border: false,
						initialZoom: 1
					}
				}}
				parent={{
					getPath: () => ({
						get: () => 0
					}),
					nodes: {
						size: 2
					}
				}}
				editor={{}}
			/>
		)

		component.instance().changeProperties({ mockProperties: 'mock value' })

		expect(Transforms.setNodes).toHaveBeenCalled()
	})

	test('IFrame opens EditIFrameModal when a src is set', () => {
		const component = renderer.create(
			<IFrame
				element={{
					content: {
						src: 'mock-src'
					}
				}}
			/>
		)
		const button = component.root.findAllByType('button')[1]
		button.props.onClick({
			preventDefault: () => {},
			stopPropagation: () => {}
		})
		expect(component.toJSON()).toMatchSnapshot()
	})

	test('IFrame opens NewIFrameModal after onChange... is pressed', () => {
		const component = renderer.create(
			<IFrame
				element={{
					content: {
						controls: '',
						border: false,
						initialZoom: 1
					}
				}}
			/>
		)
		const button = component.root.findAllByType('button')[1]
		button.props.onClick(
			{
				preventDefault: () => {},
				stopPropagation: () => {}
			},
			'mock-src-to-change'
		)
		expect(component.toJSON()).toMatchSnapshot()
	})
})
