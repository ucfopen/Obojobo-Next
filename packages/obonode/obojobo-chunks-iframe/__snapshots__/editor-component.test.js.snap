// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IFrame Editor Node IFrame component 1`] = `
<div>
  <div
    className="obojobo-draft--chunks--iframe viewer pad is-previewing  is-not-bordered is-not-showing  is-controls-enabled  is-not-missing-src is-not-scaled-up"
    contentEditable={false}
  >
    <div
      className="editor-container   is-not-selected"
      onClick={[Function]}
      style={
        Object {
          "height": "200px",
          "userSelect": "none",
        }
      }
    >
      <div
        className="obojobo-draft--components--button is-not-dangerous align-center delete-button"
        contentEditable={false}
      >
        <button
          className="button"
          contentEditable={false}
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
          type="button"
        >
          ×
        </button>
      </div>
      <div
        className="iframe-toolbar"
      >
        <span
          aria-hidden={true}
          className="title"
          contentEditable={false}
        >
          mockTitle
        </span>
        <div
          className="obojobo-draft--components--button is-not-dangerous align-center properties-button"
          contentEditable={false}
        >
          <button
            className="button"
            contentEditable={false}
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={-1}
            type="button"
          >
            IFrame Properties
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`IFrame Editor Node IFrame component handles tab 1`] = `"<div><div class=\\"obojobo-draft--chunks--iframe viewer pad is-previewing  is-not-bordered is-not-showing  is-controls-enabled  is-missing-src is-not-scaled-up\\" contenteditable=\\"false\\"><div class=\\"editor-container   is-not-selected\\" style=\\"height: 500px; user-select: none;\\"><div class=\\"obojobo-draft--components--button is-not-dangerous align-center delete-button\\" contenteditable=\\"false\\"><button type=\\"button\\" class=\\"button\\" tabindex=\\"-1\\" contenteditable=\\"false\\">×</button></div><div class=\\"iframe-toolbar\\"><span class=\\"title\\" aria-hidden=\\"true\\" contenteditable=\\"false\\">IFrame missing src attribute</span><div class=\\"obojobo-draft--components--button is-not-dangerous align-center properties-button\\" contenteditable=\\"false\\"><button type=\\"button\\" class=\\"button\\" tabindex=\\"-1\\" contenteditable=\\"false\\">IFrame Properties</button></div></div></div></div></div>"`;

exports[`IFrame Editor Node IFrame opens EditIFrameModal when a src is set 1`] = `
<div>
  <div
    className="obojobo-draft--chunks--iframe viewer pad is-previewing  is-not-bordered is-not-showing  is-controls-enabled  is-not-missing-src is-not-scaled-up"
    contentEditable={false}
  >
    <div
      className="editor-container   is-not-selected"
      onClick={[Function]}
      style={
        Object {
          "height": "500px",
          "userSelect": "none",
        }
      }
    >
      <div
        className="obojobo-draft--components--button is-not-dangerous align-center delete-button"
        contentEditable={false}
      >
        <button
          className="button"
          contentEditable={false}
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
          type="button"
        >
          ×
        </button>
      </div>
      <div
        className="iframe-toolbar"
      >
        <span
          aria-hidden={true}
          className="title"
          contentEditable={false}
        >
          mock-src
        </span>
        <div
          className="obojobo-draft--components--button is-not-dangerous align-center properties-button"
          contentEditable={false}
        >
          <button
            className="button"
            contentEditable={false}
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={-1}
            type="button"
          >
            IFrame Properties
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`IFrame Editor Node IFrame opens NewIFrameModal after onChange... is pressed 1`] = `
<div>
  <div
    className="obojobo-draft--chunks--iframe viewer pad is-previewing  is-not-bordered is-not-showing  is-controls-enabled  is-missing-src is-not-scaled-up"
    contentEditable={false}
  >
    <div
      className="editor-container   is-not-selected"
      onClick={[Function]}
      style={
        Object {
          "height": "500px",
          "userSelect": "none",
        }
      }
    >
      <div
        className="obojobo-draft--components--button is-not-dangerous align-center delete-button"
        contentEditable={false}
      >
        <button
          className="button"
          contentEditable={false}
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
          type="button"
        >
          ×
        </button>
      </div>
      <div
        className="iframe-toolbar"
      >
        <span
          aria-hidden={true}
          className="title"
          contentEditable={false}
        >
          IFrame missing src attribute
        </span>
        <div
          className="obojobo-draft--components--button is-not-dangerous align-center properties-button"
          contentEditable={false}
        >
          <button
            className="button"
            contentEditable={false}
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={-1}
            type="button"
          >
            IFrame Properties
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`IFrame Editor Node IFrame renders with no size correctly 1`] = `
<div>
  <div
    className="obojobo-draft--chunks--iframe viewer pad is-previewing  is-not-bordered is-not-showing  is-controls-enabled  is-not-missing-src is-not-scaled-up"
    contentEditable={false}
  >
    <div
      className="editor-container   is-selected"
      onClick={[Function]}
      style={
        Object {
          "height": "500px",
          "userSelect": "none",
        }
      }
    >
      <div
        className="obojobo-draft--components--button is-not-dangerous align-center delete-button"
        contentEditable={false}
      >
        <button
          className="button"
          contentEditable={false}
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
          type="button"
        >
          ×
        </button>
      </div>
      <div
        className="iframe-toolbar"
      >
        <span
          aria-hidden={true}
          className="title"
          contentEditable={false}
        >
          mockSrc
        </span>
        <div
          className="obojobo-draft--components--button is-not-dangerous align-center properties-button"
          contentEditable={false}
        >
          <button
            className="button"
            contentEditable={false}
            onClick={[Function]}
            onKeyDown={[Function]}
            tabIndex={0}
            type="button"
          >
            IFrame Properties
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
